### PR TITLE
rhsm_register: Avoid potentially checking if we need to register twice

### PR DESCRIPTION
**Pull request by tas50**
Switch to shell_out
Use include? instead of a regex
Add some simple YARD
Memoize the rhsm registration check to avoid potentially running that check twice
Stop using grep to find text. Let ruby do it all

May help with:
https://github.com/chef/chef/issues/9295

Signed-off-by: Tim Smith <tsmith@chef.io>

### CODE REVIEW

File: lib/chef/resource/rhsm_register.rb [hunk 1]
```diff
@@ -121,24 +121,30 @@ class RhsmRegister < Chef::Resource
...
 
+        #
+        # @return [Boolean] is katello-ca-consumer installed
+        #
         def registered_with_rhsm?
-          # FIXME: use shell_out
-          cmd = Mixlib::ShellOut.new("subscription-manager status", env: { LANG: "en_US" })
-          cmd.run_command
-          !cmd.stdout.match(/Overall Status: Unknown/)
+          @registered ||= !shell_out("subscription-manager status").stdout.include?("Overall Status: Unknown")
```
Comment (phiggins) [inline lib/chef/resource/rhsm_register.rb:135]: Was leaving off the LANG intentional?

File: lib/chef/resource/rhsm_register.rb [hunk 2]
```diff
@@ -121,24 +121,30 @@ class RhsmRegister < Chef::Resource
...
       action_class do
+        #
+        # @return [Symbol] dnf_package or yum_package depending on OS release
+        #
         def package_resource
           node["platform_version"].to_i >= 8 ? :dnf_package : :yum_package
         end
 
+        #
+        # @return [Boolean] is katello-ca-consumer installed
```
Comment (phiggins) [inline lib/chef/resource/rhsm_register.rb:132]: Was this supposed to have the same description as the following method?

File: lib/chef/resource/rhsm_register.rb (hunk 1 above)
Comment (tas50) [reply lib/chef/resource/rhsm_register.rb:135]: shell_out should be defaulting to en-us with the new changes

**Review by phiggins — APPROVED** [review]

File: lib/chef/resource/rhsm_register.rb (hunk 2 above)
Comment (tas50) [reply lib/chef/resource/rhsm_register.rb]: fixed

**Pull request merged**